### PR TITLE
Add CI mode

### PR DESCRIPTION
**Pull request by delner**
This pull request improves support for CI tracing by adding a "CI" mode to the tracer, which can be activated to configure settings more suitable for a test suite that runs in CI.

Users can activate CI mode via the following:

```ruby
Datadog.configure do |c|
  c.ci_mode.enabled = true
end
```

## Description

Ruby tracer_CI flow - CI_Test Mode

**Test mode:**

The tracer-level implementation, who's purpose is to configure tracing to be test friendly (capture all spans, permit diversion to buffer or alternate output.) It is not CI aware; it merely facilitates reconfiguration.

 - `c.test_mode.enabled`/`DD_TRACE_TEST_MODE_ENABLED` to enable
 - Sets `AllSampler` (100% sampler)
 - Activates `SyncWriter` (to ensure all spans are captured)
 - Provides options to configure `context_flush` or `writer_options` if desired
 
**CI mode:**

The CI-level implementation, who's purpose is to configure a tracing system to produce traces suitable for CI ingestion. It is a superset configuration to `test_mode`.
 
 - `c.ci_mode.enabled`/`DD_TRACE_CI_MODE_ENABLED` to enable
 - Activates `test_mode`
 - Adds origin tag on all spans (by activating `CI::ContextFlush::Finished`)

## Work in progress:

- [X] Fix SyncWriter/agent settings
- [X] Add :context_flush option to tracer
- [X] Refactor CI namespace
- [X] Refactor RSpec/Cucumber to common use
- [X] Add test mode settings (enabled, context_flush, writer_options)
- [X] Implement test mode
- [X] Add origin tag to RSpec/Cucumber instrumentation
- [x] Add CI::ContextFlush (origin tagging on every span)
- [x] Add CI mode setting (enabled)

## Tests:

- [ ] "test mode" Rspec context helper
- [x] "CI mode" Rspec context helper
- [X] Components
- [X] Settings
- [X] CI::Test
- [x] Origin tag on all CI spans

### CODE REVIEW

**Comment by delner** [comment]
Most of the changes should be in place. Needs tests for:

- [ ] Components
- [ ] Settings
- [ ] CI::Test
- [ ] Checks for origin tags on CI spans

**Comment by delner** [comment]
Might also want to add documentation for the `test_mode` setting... there are more changes to make for `test_mode` so I may wait until those are in (possibly in another PR)

File: lib/ddtrace/configuration/settings.rb
Comment (marcotc) [inline lib/ddtrace/configuration/settings.rb:261]: What does the default 0.5 second flush interval trying to accomplish in test mode?

Comment (delner) [reply lib/ddtrace/configuration/settings.rb:261]: It's a temporary measure adopted from https://github.com/DataDog/dd-trace-rb/pull/1385. Goal is to keep the number of spans in the buffer down, and from being dropped. (Because tests can generate a lot of spans very quickly.)

Ideally this wouldn't need to happen; we'd go to straight span streaming or sync writer. These kind of changes will be easier to accomplish after the writer refactor.

File: lib/ddtrace.rb
Comment (AdrianLC) [inline lib/ddtrace.rb:85]: There is this public repo https://github.com/DataDog/datadog-ci where we have a command but it's more used for synthetics and sourcemaps. I think that makes this a bit confusing maybe you could namespace it under `datadog/ciapp` or `ddtrace/contrib/ciapp`

Comment (delner) [reply lib/ddtrace.rb:85]: That's a JS repo. Is there an equivalent Ruby repo?

Comment (AdrianLC) [reply lib/ddtrace.rb:85]: No, there is no equivalent in ruby, but this is a CLI that can be installed by users from any language.

Comment (delner) [reply lib/ddtrace.rb:85]: If it's not in Ruby, and we don't plan to port it to Ruby, then I don't see any conflict. There's also no reason that if it were, that they couldn't be made to coexist inside the same CI namespace.

Comment (ericmustin) [reply lib/ddtrace.rb:85]: I thiink it should be ok to have `datadog/ci`, I don't see folks getting confused by this in ruby.

File: lib/ddtrace/ext/test.rb
Comment (ericmustin) [inline lib/ddtrace/ext/test.rb:5]: Is this consistent across languages?

File: lib/ddtrace/configuration/settings.rb
Comment (ericmustin) [inline lib/ddtrace/configuration/settings.rb:249]: I am pretty confused on when a user would use this setting, and what environments (it seems like there's a list of vendors it supports?). But I assume this will be addressed in documentation at some point?

File: lib/ddtrace/configuration/components.rb
Comment (ericmustin) [inline lib/ddtrace/configuration/components.rb:123]: is the side effect because the root span in each flush gets assigned origin and sampling priority?

File: lib/ddtrace/configuration/components.rb
Comment (ericmustin) [inline lib/ddtrace/configuration/components.rb:127]: In these test environments, is the trace-agent running in [syncMode](https://github.com/DataDog/datadog-agent/blob/140a4ee164261ef2245340c50371ba989fbeb038/pkg/trace/writer/trace.go#L65) as well, ala how it runs in Lambda? I do think that sync writing makes sense to work toward if we need high guarantees that all spans flush.

**Review by ericmustin — COMMENTED** [review]
Left a few comments and questions that are not blocking. It was a bit hard to tell what's new code bc of the `CI` refactor, but I generally think this approach with forcing "span streaming" is a nifty way accomplish adding origin to every span. Not sure about the practicality of this approach outside of a test environment, but as users won't be enabling outside of that environment, then it's fine with me. Working to a sync writer is a good goal.

I also think this work is a good indication that we'll need to anticipate future use cases of trace or transaction level metadata being attached to every exported span. Although, again, I understand the rfc/practicality limits what we can do.

looks like a few tasks open on PR but I think this is good to approve otherwise

File: lib/ddtrace/ext/test.rb
Comment (delner) [reply lib/ddtrace/ext/test.rb:5]: No, this is an original creation as far as I'm aware.

File: lib/ddtrace/configuration/settings.rb
Comment (delner) [reply lib/ddtrace/configuration/settings.rb:249]: Yeah, it has to be documented. Presumably you'd set this mode in your test suite if you want to tune the tracer to be more test friendly (e.g. keep 100% traces, produce traces to a test buffer, etc)

We've had users ask for a test mode for a long time in Ruby. https://github.com/DataDog/dd-trace-rb/pull/187

File: lib/ddtrace/configuration/components.rb
Comment (delner) [reply lib/ddtrace/configuration/components.rb:123]: It's a side effect because the "partial flush" always tags the top most span with origin and sampling priority. We actually do want origin on every span for tests, I don't think it matters if priority is there or not.

File: lib/ddtrace/configuration/components.rb
Comment (delner) [reply lib/ddtrace/configuration/components.rb:127]: Right now, no. But it could, potentially.

Downside with synchronous writing is that it doesn't play nicely with some anticipated use.

I would like users to be able to assert things about traces produced within their tests (e.g. "I expected this web request to produce a trace with a Rack span") This would be especially useful for users who have defined manual instrumentation and want to verify its correctness.

If users use the RSpec or Cucumber integration to trace their CI tests, the root span of their traces will be the span for the test (e.g. RSpec span.) Because they would be running assertions on trace output while the test is still running, the trace would not be complete and not written to the writer's trace buffer. Any assertions on trace output will fail, because the trace will not be in the buffer to assert against.

This is where partial flush of size 1 comes in: it would "stream" each span from the context immediately to the writer's buffer, meaning you could assert on written spans. It also has the bonus of tagging every span with "origin", something we currently require of all spans generated during a CI test.

Sync writing doesn't play well with this, because if we activated it with partial flushing size 1, it would do a blocking HTTP request for every span in the test... this would be very bad.

There may be a better strategy to get these two to play together, but I don't have one at the moment. If you have suggestions I'd be happy to hear them.

**Comment by delner** [comment]
@ericmustin Sounds like we're pretty much on the same page.

If it helps to review the code, you could look at the commits instead of the full list of changes: the commits after the refactor are pretty small.

File: lib/ddtrace/ext/test.rb
Comment (ericmustin) [reply lib/ddtrace/ext/test.rb:5]: that's fine. Do we think this 'test mode' will require a sdk-specific env var across langs in the future, or is ruby the only lang that has a special mode/settings needed in these envs? Would it make sense to namespace `DD_RUBY_TEST_MODE_ENABLED` to avoid namespace clash in the future or is that overkill? fine either way imo.

File: lib/ddtrace/configuration/settings.rb
Comment (ericmustin) [reply lib/ddtrace/configuration/settings.rb:249]: kk. i think as long as we're clear this will be documented in the future this can be resolved.

File: lib/ddtrace/configuration/components.rb
Comment (ericmustin) [reply lib/ddtrace/configuration/components.rb:127]: Yes i think sync mode would mean moving away from how we currently do partial flushing, we would instead be queuing many traces/spans in memory (perhaps >10k), and then doing a few large flushes. I would have to review how the agent handles syncronous flushing, ideally we could model our approach after theirs.

Comment (delner) [reply lib/ddtrace/configuration/components.rb:127]: I think this might be worth engaging with more, now, if there's some reasonable, effective change we can make. I'd rather not mess with buffer/interval settings and expose a public API for it if it isn't comprehensive and there's a better alternative readily available.

I will consider alternate strategies, and share my findings.

File: lib/ddtrace/ext/test.rb
Comment (delner) [reply lib/ddtrace/ext/test.rb:5]: It's really hard to say... if I had to guess, I don't think all languages will have a need for this: it's most likely informed by how each tracer operates between production and test environments, and the accessibility to changing tracer function to suit these needs. I can see the possibility that some language doesn't have such a need, and I could potentially see a future version of the Ruby tracer that doesn't need it either.

I don't think namespacing by language makes sense, but it may be better to namespace by "product" e.g. `DD_TRACE_TEST_MODE_ENABLED`, so if that seems sensible to you, I'll make that change.

**Comment by delner** [comment]
Updated my TODO list per review feedback; see original comment.

File: lib/ddtrace/configuration/settings.rb
Comment (delner) [reply lib/ddtrace/configuration/settings.rb]: I dropped this, and implemented `SyncWriter` instead.

File: lib/ddtrace/ext/test.rb
Comment (delner) [reply lib/ddtrace/ext/test.rb]: Changed this.

File: lib/ddtrace/configuration/components.rb
Comment (delner) [reply lib/ddtrace/configuration/components.rb]: No longer relevant.

File: lib/ddtrace/configuration/components.rb
Comment (delner) [reply lib/ddtrace/configuration/components.rb]: I decided to go a different direction, based on the feedback.

By default, CI will turn on a SyncWriter & a custom context flush (same as the default, but adds tags.) Partial flushing no longer applies. This should help improve the guarantee that all spans are flushed without having conflicts with partial flush.

Let me know what you think.

**Review by marcotc — APPROVED** [review]
I like this new approach! 🎉

**Review by ericmustin — COMMENTED** [review]
I think the introduction of an explicit `ci_mode` is appropriate here. I will be happy to approve once implemented and we ensure that `origin` is adding to every span when this mode is enabled

File: lib/ddtrace/configuration/components.rb
Comment (ericmustin) [reply lib/ddtrace/configuration/components.rb]: I think this makes sense. Am I right in saying that the `context_flush` bit about adding tags is not yet implemented?

I think a combination of `context_flush` + `SyncWriter` are correct here. The `test_mode` i'd be curious to see whether users find it to be useful, i know there have been a lot of requests for it, so that's a nice bonus we can get out of this work.

Comment (ericmustin) [reply lib/ddtrace/configuration/components.rb]: I think this comment thread can be resolved when you feel appropriate

**Comment by delner** [comment]
@ericmustin I've added a spec to RSpec and Cucumber that activates this mode and asserts the presence of this tag on every span. That should be the test that we need.

@ericmustin @marcotc should be ready for a final review (might need to get CI passing though, too.)

**Review by marcotc — APPROVED** [review]
I like the separation of CI components and configuration to their own files. 👍 

The test failures seem related to this PR (not just flaky tests), so I'll approve the changes, but we'll probably need one more ➕ before it's finally merged.

**Review by ericmustin — APPROVED** [review]
I haven't looked too closely at the 3.0 failures, but setting those aside, some small Qs but i think this lgtm. awesome work.

Not blocking, but i do want to re-iterate that we may want to think about how to extend `ContextFlush::Tagging.get_trace` in the future. Currently we add `origin` as a tag, but I envision a scenario in the future where we will need additional, arbitrary, vendor specific tags appended to every span at export/encode time (trace or transaction -level tags). we may want to think about `origin` as simply a particular tag name within a more inclusive `context.trace_state` field, which is a w3c concept https://w3c.github.io/trace-context/#tracestate-header. Instead of `attach_origin` we could simply expose a method like `attach_trace_state(span, trace_state_tag_name)` or `attach_all_trace_state(span)`. 

Don't want to over-engineer for a future hypothetical, but i anticipate requests like this becoming more regular, so just offering this for posterity/future reference if this becomes a pain point

File: lib/datadog/ci/context_flush.rb
Comment (delner) [reply lib/datadog/ci/context_flush.rb:12]: That was not in our spec for CI, so no, I don't think so. If we have to, that should be pretty easy to do.

**Comment by delner** [comment]
@ericmustin I really like the idea for "attach trace state"; my recommendation is the next time we have to decorate more state that we seriously considering implementing this pattern. May be premature for the first pass, but justified for the second.

File: lib/datadog/ci/context_flush.rb
Comment (ericmustin) [reply lib/datadog/ci/context_flush.rb:12]: 👍

**Review by marcotc — APPROVED** [review]

**Review by ericmustin — APPROVED** [review]

**Comment by marcotc** [comment]
As a reminder to ourselves, this is how one configures CI tracing:
```ruby
# spec_helper.rb
require 'datadog/ci'# Enable CI tracing
Datadog.configure do |c|
  c.ci_mode.enabled = true
  c.use :rspec
end
```

**Pull request merged**
